lib/test-utils/src/Test/Hspec/Extra.hs

```diff
@@ -126,28 +135,34 @@ itWithCustomTimeout
     -> String
     -> ActionWith ctx
     -> SpecWith ctx
-itWithCustomTimeout sec title action = specify title $ \ctx -> timeout sec $ do
-    action ctx
-        `catch` (\(_ :: AsyncCancelled) -> return ())
-        `catch` (\(e :: SomeException)  -> action ctx
-        `catch` (\(_ :: SomeException)  -> throwIO e))
+itWithCustomTimeout sec title action = specify title $ \ctx -> do
+    e <- newEmptyMVar
+    race (timer >> tryTakeMVar e) (action' (void . tryPutMVar e) ctx) >>= \case
+       Left Nothing ->
+           assertFailure $ "timed out in " <> show sec <> " seconds"
+       Left (Just firstException) ->
+           throwIO firstException
+       Right () ->
+           pure ()
```

Comment:
Hmmmm. Okay. We are loosing the "AsyncCancelled" handler and tests will report as failed when we CTRL-C them but It seems to be for the best.

Reply:
This is one of the nice things about unliftio - less things to think about when handling exceptions.
`UnliftIO.Exception.catch` will only catch synchronous exceptions.
Any `AsyncCancelled` or `ThreadKilled` exceptions can bubble up.
The disadvantage is that an MVar is needed here since AsyncCancelled (end of race) is never caught.